### ISSUE TITLE
Performance work

### ISSUE THREAD

**Issue by huitema** [private-octopus/picoquic#78]
The current tests indicate a sustainable rate of about 400 Mbps through a simulator. We need to build an actual performance test, with two variants: high speed data transfer; and, high rate of transactions.

**Comment by huitema** [comment]
Possible work include moving to a zero copy architecture, and threading.

**Comment by bkchr** [comment]
Hi, did you have any working benchmark for picoquic? And the simulator, what did that simulate? I'm currently want to create some sort of video streaming in my local network, but I takes around 500ms to submit one image (50-60kb). I don't know if I'm doing something wrong on my application side or what else is happening. My network shows a constant upload rate of 150-180kb/s and that matches the 2 images per second. Doing the transfer in local host it takes around 40ms to transfer one image.
Maybe I'm doing something completely wrong, but are there some parameters that I could tune or that I could observe to get a better understanding what I may doing wrong?

**Comment by huitema** [comment]
I need to write a wiki page on the simulator, and on the test strategy.

The simulator that I added to the test tools just does basic single link simulation. It will manage a queue, the size of which can be specified, and overflow will cause losses. In the tests I use two links, one for each direction. The tests work in "virtual time". 

I do not have any explicit performance test yet, and it is time to add them. We can start with something basic, maybe one that mimic the video streaming, simulate something like sending and receiving 100 frames, and verifies that the virtual time to do that is smaller than some target. This will help us finding basic issues like synchronization, bad timers, retransmission, flow control, etc.

There are other scenarios that we can add later, but starting with one will be better than 0.

**Comment by huitema** [comment]
I quickly added a "completion time limit" parameter to the "tls_api_one_scenario_test()" API. This is only used now for the "long transmission with congestion" test, which transfers 1MB of data over 10 Mbps links, 10 ms latency, and a maximum queue length equivalent to 10ms. The max execution time is currently set to 2 seconds -- the actual execution is 1.9 second. This is way above the expected time of 800 milliseconds + some overhead, and I will start a specific issue on this point.

The 10 Mbps bit rate and 10 millisecond latency are fixed parameters in all the current tests. We should consider making that programmable, so we can test performance in a variety of scenario.

**Comment by bkchr** [comment]
Thanks for your fast and good answers!
Is there some parameter that stores the observed loss?

**Comment by huitema** [comment]
The connection context includes counters: nb_retransmission_total is the total number of retransmission over the duration of the connection, and nb_spurious the number of spurious retransmissions.

The test tools are good for doing "non regression" testing, but they are not designed to do performance tuning. This requires simulating a scenario, collecting a time line of sequence numbers, packets sent, retransmissions, acks, congestion window changes, flow control changes, etc. Then, use some graphing tools to visualize the data and understand what is happening. This requires a specific tool. But I am travelling to ICANN this week, and to IETF next week, so I don't think I can develop that before early April.

**Comment by huitema** [comment]
Surprise. The simulated transmission time for the same scenario are longer on Linux than on Windows, e.g. 1.5 second instead of 1 second for a simple 1MB transmission. This is weird. Will need to investigate!

**Comment by bkchr** [comment]
Ohh yeah, that is really weird. I will also try to find some time to look into this.

**Comment by promgamer** [comment]
Do you have performance issues with picoquic? I've tried the demo, compiled in release mode, stripped out logging and stuff, and I can't go over ~266KB/s (it takes me +/- 15 sec to download a 3.4MB file). Tested with quant and winquic implementations.

**Comment by promgamer** [comment]
About last comment: shame on me, it turns out it is caused by packet loss on my network. Changed from wi-fi to cable, and it improved a lot (I got ~4MB per second, which is almost max-speed for my current internet connection).

Thank you @huitema for your help!

**Comment by huitema** [comment]
Issue #463, #467 and #469 are performance issues, part of the CPU performance work. Other issues will include threading.

**Comment by huitema** [comment]
PR #465 include zero-copy on send, which fixes one of the potential performance issues. We may need something similar on receive.

**Comment by huitema** [comment]
Measurements by @hoang-tranviet on a 1 Gbps local network show a throughput of about 250 Mbps (see comments on PR #522). This is not too bad for a prototype, but we should set an initial performance goal of achieving 1Gbps.

**Comment by huitema** [comment]
Analysis of "very long" unit test shows that over 80% of the CPU time is tied to crypto, including 17% for processing incoming Initial packets, 16% for setting the client context, and 48% for AES_GCM.

**Comment by hoang-tranviet** [comment]
One thing may impact the performance is the printf() which imposes some overhead even when we direct to /dev/null.
Creating a wrapper for printf() to disable it on-demand or even support different levels may be a way.
Just my two cents.

**Comment by huitema** [comment]
Yes, controlling the log level would help. Do you have measurements?

**Comment by hoang-tranviet** [comment]
No, I have not implemented it yet. Now I am looking at some hacking way to override printf() instead.

**Comment by huitema** [comment]
So I just started doing performance testing on Windows. Basic test in debug mode, client to server over loopback socket, measure 99 Mbps. Code analysis shows that 66% of CPU time is spent in "fprintf", i.e. on the log output, which is very much the issue that @hoang-tranviet was noticing. As a consequence, I put up PR #558, which allows for a "no log at all" option, and also by default stops logging after 100 packets have been sent. With this changes, the cost of "fprintf" drops to less than 1.5% of the running cost when log is enabled, and 0% when it is not.

**Comment by bkchr** [comment]
And did the performance increase?

**Comment by huitema** [comment]
Removing the logs basically doubles the performance. At that point, throughput reaches somewhat over 300 Mbps when sending 10MB over the loopback connection on my Windows laptop. Interestingly, the throughput drops to half that if sending 100MB instead of 10MB, which probably points to a congestion control issue. On windows, most of the time (55 to 60%) is spent in the "WSASendMsg" code, and less than 10% is spent on encryption decryption. The rest is spent in the various parsing and composition tasks.

**Comment by huitema** [comment]
On a Linux VM on the same laptop, I measure 480 Mbps when sending 10MB over loopback, but 250 Mbps when sending 100MB. Without checking the details, it would seem that UDP loopback works significantly better on Linux than on Windows.

**Comment by huitema** [comment]
Of course, If I enable congestion control traces (using `picoquic_set_cc_log`) the throughput drops, because the logs where published using `fprintf`, and logging uses 40% of the CPU. Changing that to use binary format instead, contains the log overhead to 2.5% of CPU. 

Early traces show that there is indeed congestion happening on the loopback socket. It seems that cubic is not optimized for that scenario. The congestion windows spikes to a high value during slow start at the beginning of the connection, then is reduced after a succession of packet losses, and does not grow further.

**Comment by huitema** [comment]
Check PR #559 for the new version of the congestion log, plus the new utility picoquic_c2csv that converts CC logs from binary format to CSV.

**Comment by huitema** [comment]
Found the reason between the differences in performance. Basically, Cubic is way less aggressive than Reno when the delay is small. Cubic uses a time constant `K` that is a function of the Window size:

|---------|----------|
| W(packets)| K(seconds)|
|---------|----------|
| 10 | 1.46 |
| 20 | 1.85 |
| 50 | 2.5 |
| 100 | 3.15 |

In the loopback test, the window reaches a few hundred packets. After a loss, Cubic would shrink the window, then grow it slowly for 3 seconds -- about 1000 RTT! In that time, Reno would have grown the Window by 1000 packets.

In theory, this is compensated by computing an approximation of the New Reno congestion window, and using it if it is larger than the Cubic window. But the formula is a complicated approximation that relies on RTT estimates -- and the estimate is going to fluctuate continuously during the 1000 RTT. Instead, it is much simpler to just run a parallel estimate for new Reno, and keep the highest. With that, the performance remains good even for long files. (See PR #560 for the changes.)

**Comment by huitema** [comment]
And with this change, we get the following performances on the loopback test, downloading 200 MB, on Windows, and the on an Ubuntu VM on the same laptop:

| System | Variant | Mbps |
| ------- | ------- | ----: |
| Windows | Debug | 233 |
| Windows | Release | 316 |
| Ubuntu |  | 603 |

**Comment by huitema** [comment]
Did more work  managing a packet pool, adding a "no_disk" option to the client, and reducing the impact of the calls to `picoquic_is_cnx_backlog_empty`. With that we get the following:

System | Variant | Mbps
-- | -- | --
Windows | Debug | 269
Windows | Release | 521
Ubuntu |   | 653

At this point, the main resource drain on Windows appears to be the calls to UDP sockets "msgsend", "msgrecv" and "select". The socket code consumes 35% of the client time, 50% of the server time.

**Comment by aneesh-joshi** [comment]
@huitema @bkchr 
Could you share the script for these performance benchmarks?
I am interested in getting some numbers for a simple file transfer over QUIC vs TCP.

**Comment by huitema** [comment]
@aneesh-joshi The perfomance tests were done on a loopback connection, turning off the logging facility on server and client, and instructing the client to not write to disk. Server and clients were launched in different terminals. The commands would be:
```
On server:
picoquicdemo -p 4433

On client:
picoquicdemo -D ::1 4433 /200000000
```
On windows, use `picoquicdemo.exe` instead of `picoquicdemo`.

You can do the same tests over the network, e.g. download a 10MB file as:
```
picoquicdemo -D test.privateoctopus.com 4433 /10000000
```
By default, the server will only log the first 100 packets of the connection.

The same 10MB file can be downloaded from the Apache2 server on the test machine as:
```
wget https://test.privateoctopus.com/10000000
```

**Comment by aneesh-joshi** [comment]
Is there any way I can give it a custom file to transfer between a quic server and a quic client?

Ideally, I would just like something as simple as:
1. Create a socket
2. listen on it
3. accept a connection
4. read the data

This way, I can keep the benchamark pretty similar between QUIC and TCP. But there seems to be a lot of other stuff going on in the `picoquicdemo`

Any thoughts or help will be appreciated.

**Comment by aneesh-joshi** [comment]
Hey, @huitema 
I'm trying to test this over a network.
What's the correct way of specifying the IP.

Currently, I'm trying:

Server:
`./picoquicdemo -p 4433`

Client:
`./picoquicdemo  0:0:0:0:0:ffff:c0a8:102 4433 /200000000`

I get the following output at client:

**Comment by huitema** [comment]
@aneesh-joshi
There are only 6 components in the IPv6 address that you specified. Are you sure that you specified it correctly?

**Comment by aneesh-joshi** [comment]
I think I incorrectly translated my ipv4 to the ipv6 without realising that the interface needs to be registered with the ipv6 address.

I tried it with the ipv4 address and it doesn't work either. I suspect the server is only listening on the loopback.

Could you suggest commands to get a simple client server transfer?

**Comment by huitema** [comment]
You should start the server as:
`picoquicdemo -l log.txt -p 4433`
That will create a log in the specified file. 

You should also start the client as
`picoquicdemo -l client_log.txt server_name 4433 /5000000`
That will give you a client log. It should make debugging easier. Of course, you need to replace the server name by the actual name of the server, or its IP address.

**Comment by basu13** [comment]
Will you provide any sample code for client and server file transfer in c++/c using picoquic. I am successfully able to build it but I don,t how to use it

**Comment by huitema** [comment]
@iriszero After implementing UDP GSO on Linux, and the equivalent on Windows, the performances (using Cubic) are now:

System | Variant | Mbps
-- | -- | --
Windows | Debug | 1,048
Windows | Release | 1,605
Ubuntu |   | 1,300

On Windows Release, it appears that over 76% of the time is spent in `picoquic_sendmsg_start`.

**Comment by huitema** [comment]
In a comment on PR #1113 , @tbarbette wrote: _Fyi now an 8-core 3Ghz Xeon can do around 10Gbps both for client and server (HTTP "0.9", serving the generated file /10000, pretty convenient btw). The biggest issue is I/O. We could use DPDK and/or AF_XDP to mostly solve that, but in our case, we just wanted to have some QUIC traffic to design a load-balancer.
The server simply launched one picoquicdemo server per core, using network namespaces to separate them. SRIOV can ensure packets go to the right core in hardware so it's quite efficient._

**Comment by huitema** [comment]
We now have performance data from the "perf" interop before IETF 110.

Upload performance on Windows:

↓Client   →Server | msquic | picoquic | lsquic | ngtcp2 | quinn
-- | -- | -- | -- | -- | --
msquic | 2.23 Gbps | 4.57 Gbps | X | X | X
picoquic | 2.12 Gbps | 2.01 Gbps | X | X | 1.36 Gbps
lsquic | X | X | X | X | X
quinn | X | X | X | X | X

Upload performance on Ubuntu 20.04

↓Client   →Server | msquic | picoquic | lsquic | ngtcp2 | quinn 
-- | -- | -- | -- | -- | -- 
msquic | 2.75 Gbps | 3.45 Gbps | 3.09 Gbps | 3.77 Gpbs | X 
picoquic | 1.24 Gbps | 1.26 Gbps | 1.30 Gbps | 1.17 Gbps | X 
lsquic | 2.43 Gbps | 0.97 Gbps | 2.17 Gbps | 2.52 Gbps | X 
quinn | X | X | X | X | X 

Download performance on Windows

↓Client   →Server | msquic | picoquic | lsquic | ngtcp2 | quinn 
-- | -- | -- | -- | -- | -- 
msquic | 2.23 Gbps | 1.99 Gbps | X | X | X 
picoquic | 2.12 Gbps | 2.01 Gbps | X | X | X 
lsquic | X | X | X | X | X 
quinn | X | X | X | X | X 

Download performance on Ubuntu 20.04

↓Client   →Server | msquic | picoquic | lsquic | ngtcp2 | quinn  
-- | -- | -- | -- | -- | --
msquic | 2.75 Gbps | 3.45 Gbps | 3.09 Gbps | 3.77 Gpbs
picoquic | 1.24 Gbps | 1.26 Gbps | 1.30 Gbps | 1.17 Gbps
lsquic | 2.43 Gbps | 0.97 Gbps | 2.17 Gbps | 2.52 Gbps
quinn |   |   |   |     

Replies per second (RPS), 250 Connections, 0 byte Request, 4096 Byte Response, Windows 10

↓Client →Server | msquic | picoquic | lsquic | ngtcp2 | quinn  
-- | -- | -- | -- | -- | -- 
msquic | 547K RPS | 15K RPS | X | X |
picoquic | X | X | X | X  |
lsquic | X | X | X | X |   |  
quinn |   |   |   |   |   |  

Parallel Handshakes per Second (PHS), Windows   10

↓Client →Server | msquic | picoquic | lsquic | ngtcp2 | quinn
-- | -- | -- | -- | -- | -- 
msquic | 2640 HPS | 518 HPS | X | X |   
picoquic | X | X | X | X |     
lsquic | X | X | X | X |   
quinn |   |   |   |   |   

This was done with a test bed provided by the msquic team at Microsoft. The test client was connected to the server by a 10 Gbps link, which TCP connections could saturate -- check, this probably involved hardware offload. 

Key lessons:
* Picoquic server reaches best or close to best upload and download performance when running against the Msquic client. This is very good news for server performance, not so much for client performance. The picoquic client in this test had a bug in support of delayed ACK, that has now be fixed. Need to verify that the client reaches performance parity with msquic in tests against picoquic server.
* The RPS test shows much worse performance than msquic. Need to understand that. Is it due to some kind of flow control effect? Or maybe the handling of parallel connections? Need to understand that.
* The PHS test shows worse performance than msquic. Need to understand that. Is it due to using an RSA certificate for test? Would using an EC-DSA certificate change that? Should we work on that part of the code?

**Comment by huitema** [comment]
Closing. Performances are OK, although the client should probably be improved.
